### PR TITLE
[top, util] Add quick support for multiple power domain resets

### PR DESCRIPTION
**Pull request by tjaychen**
This commits adds the ability for modules to be declared under
multiple power domains and thus assign resets of different domains
to each port.  This is done by adding more information to the `reset_connections`
field.  Now each connected reset has both a name and a related domain. 

To avoid changing existing configurations, modules with only one domain specified (which is all of them)
will automatically be converted as part of the validation process.  Those that have multiple
domains will error if the user does not explicitly choose one to be related to the reset. 

This change is done entirely to support AST's need for multiple domains underneath, which is different
from all comportable IPs (single domain only). See #7979 

**WARNING**
Note this is a quick bolt on and not the ideal way of doing things.
Ideally, the concept of multiple domains should be expressed in the
ip.hjson themselves. The top level then hierarchically connects the
available top level domains to the sub-domains.

The ip.hjson would then also clearly state what local domain each
reset is related to, and the connective tissue should be built-up that
way instead of being mandated from the top.

However, that is a larger change than necessary at this point, and should
be accounted for later in the reset generation refactoring.

Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
@zi-v 
@arnonsha

**Review by msfschaffner — APPROVED** [review]
Should we open a bug to track potential future improvements of this mechanism?

**Comment by tjaychen** [comment]
> Should we open a bug to track potential future improvements of this mechanism?

yeah there's one already, let me link to this PR also.

File: util/topgen/validate.py [hunk 1]
```diff
@@ -653,6 +660,37 @@ def validate_reset(top, inst, reset_nets, prefix=""):
     log.info("%s %s resets are %s" %
              (prefix, name, reset_signals))
 
+    # Check if reset connections are properly formatted
+    # There are two options
+    # The reset connection for a particular port must be a str
+    # The reset connection for a paritcular port must be a dict
+    # If value is a string, the module can only one have ONE domain
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply util/topgen/validate.py:667]: fixed!

File: util/topgen/validate.py [hunk 2]
```diff
@@ -741,29 +779,23 @@ def check_flash(top):
...
-                      .format(end_point['name'],
-                              end_point['domain']))
-            error += 1
-            return error
-
-    # arrived without incident, return
-    return error
+        for d in end_point['domain']:
+            if d not in top['power']['domains']:
+                raise ValueError("{} defined invalid domain {}"
```
[the remark this answers is not in the document]
Comment (tjaychen) [reply util/topgen/validate.py]: fixed!

**Pull request merged**